akka-stream compiler warnings as fatal errors

johanandren opened the pull request:
Refs #26088

Scala/javadoc validation failing, waiting for #26634 to get that disabled

johanandren committed:
Lots of warnings eliminated, stuck on non-nullary method overrides nullary method

raboof commented:
#26634 is merged so this can be rebased

raboof commented on akka-stream/src/main/scala/akka/stream/impl/streamref/StreamRefsMaster.scala:
```diff
@@ -23,9 +23,7 @@ private[stream] object StreamRefsMaster extends ExtensionId[StreamRefsMaster] wi
-private[stream] final class StreamRefsMaster(system: ExtendedActorSystem) extends Extension {
-
-  private val log = Logging(system, getClass)
+private[stream] final class StreamRefsMaster(@unused system: ExtendedActorSystem) extends Extension {
```
Since it's private, can we remove the `system` parameter?

johanandren replied:
I had some bad luck when I was thinking about how extensions are instantiatied, yes, we should be able to remove that.

johanandren commented on akka-stream/src/main/scala/akka/stream/Transformer.scala:
```diff
@@ -4,6 +4,8 @@
+
 import scala.collection.immutable
 
 private[akka] abstract class TransformerLike[-T, +U] {
```
But it's such a cool class name ;(

johanandren committed:
Unused stuff eliminated

johanandren committed:
Type parameter added in internals but not in test

johanandren committed:
Mima filters

johanandren committed:
Messed up the plugin config in the rebase, fixed

johanandren committed:
Disable warnings and fail warnings for Scala 2.11

Or else we are stuck on a nullary overriding non nullary with no source file reference

patriknw commented on akka-stream/src/main/scala/akka/stream/impl/UnfoldResourceSourceAsync.scala:
```diff
@@ -107,11 +107,12 @@ import scala.util.control.NonFatal
-              case Success(r)  => close(r)
+              case Success(r) =>
+                close(r)
               case Failure(ex) => throw ex // failed to open but stream is stopped already
```
where does this exception go? do we care about it?

johanandren replied:
Default exception handler on the dispatcher threadpool?

patriknw commented on akka-stream/src/main/scala/akka/stream/javadsl/Tcp.scala:
```diff
@@ -281,7 +281,6 @@ class Tcp(system: ExtendedActorSystem) extends akka.actor.Extension {
       negotiateNewSession: NegotiateNewSession,
       backlog: Int,
       options: JIterable[SocketOption],
-      halfClose: Boolean,
```
Even if it's ApiMayChange we shouldn't just change if we don't have a strong reason. Could be changed when we remove the ApiMayChange if `halfClose` was a mistake.

johanandren replied:
I thought that the fact it wasn't used at all inside the method is readon enough?

raboof replied:
Hmm if it's not used then that is suspicious though... shouldn't it be used?

johanandren replied:
Ah, so the inconsistency is that the underlying API has it, but the Scala API does not, yes, maybe it should rather be added there and passed along there.

patriknw commented on akka-stream/src/main/scala/akka/stream/scaladsl/Tcp.scala:
```diff
@@ -117,7 +119,7 @@ final class Tcp(system: ExtendedActorSystem) extends akka.actor.Extension {
       port: Int,
       backlog: Int = 100,
-      options: immutable.Traversable[SocketOption] = Nil,
+      options: immutable.Iterable[SocketOption] = Nil,
```
API changes like this makes me super scared

johanandren replied:
Yeah, I realise maybe we can't do this. 

Iterable extends Traversable, but the other way around, it could be possible (but perhaps not likely) that a user had some special Traversable impl that wasn't Iterable.

In general they'd likely didn't specify anything or put a standard collection in here and then I think it'd be safe.

johanandren replied:
The reason for the change was that `Traversable` is deprecated in 2.13

patriknw replied:
it might be source compatible but not binary compatible

raboof replied:
There is a slightly similar situation with `IterableOnce`, which did not exist at all in 2.12 and replaces `TraversableOnce` in 2.13.

That one is worked around without binary incompatibility by defining a `type IterableOnce[+X] c.TraversableOnce[X]` in `akka.util.ccompat` for versions lower than 2.13 and using that type there, while not defining one in ccompat and using the 'real' `IterableOnce` on 2.13.

I didn't do the same for `immutable.Iterable` because unlike `IterableOnce`, `immutable.Iterable` does exist on 2.12 so that could get pretty confusing... in theory we could introduce a third name that means `Traversable` on 2.12 and `Iterable` on 2.13, but that still seems weird, might be better to just `@silent` this API?

johanandren replied:
Yeah, I'll revert and try to silence instead.

patriknw reviewed (commented):
why API changes?

johanandren committed:
Go back on the API changes

patriknw commented on akka-stream/src/main/scala/akka/stream/javadsl/Tcp.scala:
```diff
@@ -283,11 +283,10 @@ class Tcp(system: ExtendedActorSystem) extends akka.actor.Extension {
-        .map(new IncomingConnection(_))
-        .mapMaterializedValue(_.map(new ServerBinding(_))(ec).toJava))
+    Source.fromGraph(delegate
+      .bindTls(interface, port, sslContext, negotiateNewSession, backlog, immutableSeq(options), halfClose, idleTimeout)
```
Can you create an issue (and a FIXME close to the ApiMayChange here) that we should look into removing that `halfClose` parameter. It doesn't exist in scaladsl (and there it's always set to true). It's probably correct that it should always be true for TLS.

Therefore we should also not change the behavior here by passing on the parameter. Could become an issue if someone is using `false`.

johanandren replied:
Created  #26689

johanandren replied:
And reverted change

patriknw commented on akka-stream/src/main/scala/akka/stream/scaladsl/Tcp.scala:
```diff
@@ -307,12 +314,36 @@ final class Tcp(system: ExtendedActorSystem) extends akka.actor.Extension {
+      @silent // Traversable deprecated in 2.13
+      options: immutable.Traversable[SocketOption],
+      halfClose: Boolean,
+      idleTimeout: Duration): Source[IncomingConnection, Future[ServerBinding]] = {
```
are we sure this variant is needed for TLS? let's make that decision in the other issue I requested ^

patriknw commented on akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala:
```diff
@@ -705,9 +706,9 @@ abstract class GraphStageLogic private[stream] (val inCount: Int, val outCount:
-      } else andThen(result)
+          if (pos == n) andThen(result.toIndexedSeq)
+        }, () => onClose(result.take(pos).toIndexedSeq)))
+      } else andThen(result.toIndexedSeq)
```
This could be a performance regression, creating Vector from the Array each time.

johanandren replied:
Ah yes, should be `a.toSeq`, not `toIndexedSeq` to be identical to the previous implicit conversion, I'll change

johanandren replied:
Or do you think the array wrapper alloc will be troublesome as well? The wrapping should only happen at the end of the n read elements, not for each, but ofc for each call to `readN` it will happen once

johanandren replied:
See my comment to raboof below about this: https://github.com/akka/akka/pull/26673#discussion_r272530721

johanandren committed:
Review feedback

johanandren committed:
Whatt. Formatting

raboof commented on akka-stream/src/main/scala/akka/stream/javadsl/Tcp.scala:
```diff
@@ -281,6 +284,7 @@ class Tcp(system: ExtendedActorSystem) extends akka.actor.Extension {
       negotiateNewSession: NegotiateNewSession,
       backlog: Int,
       options: JIterable[SocketOption],
+      @silent // FIXME unused #26689
```
:+1:

raboof commented on akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala:
```diff
@@ -705,9 +706,9 @@ abstract class GraphStageLogic private[stream] (val inCount: Int, val outCount:
-      } else andThen(result)
+          if (pos == n) andThen(result.toSeq)
+        }, () => onClose(result.take(pos).toSeq)))
+      } else andThen(result.toSeq)
```
Not trivial to see that this is harmless from this diff, but if you verified then :+1:

johanandren replied:
Without this explicit conversion there was an implicit conversion `Predef.genericWrapArray` which calls `WrappedArray.make(array)`  which then constructs a `WrappedArray` so there is actually no added alloc overhead, the overhead was already there (checked in 2.12).

raboof commented on akka-stream/src/main/scala/com/typesafe/sslconfig/akka/SSLEngineConfigurator.scala:
```diff
@@ -21,6 +21,7 @@ final class DefaultSSLEngineConfigurator(
     enabledProtocols: Array[String],
     enabledCipherSuites: Array[String])
     extends SSLEngineConfigurator {
+  config ne null // @unused unavailable
```
adding `@unused` on the variable didn't work here? OK though.

johanandren replied:
Yea, not available because this is not in `_root_.akka`

raboof commented on project/AkkaDisciplinePlugin.scala:
```diff
@@ -64,11 +65,6 @@ object AkkaDisciplinePlugin extends AutoPlugin with ScalafixSupport {
               "-Yno-adapted-args")
           case Some((2, 12)) =>
             disciplineScalacOptions
-          case Some((2, 11)) =>
```
:+1:

chbatey reviewed (approved):
LGTM, very silent

chbatey commented:
I'll let you merge this first @johanandren and i'll deal with the conflict in AKkaDiscipline
